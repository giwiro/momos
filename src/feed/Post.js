// @flow
import React, {Component} from 'react';
import PropTypes from 'prop-types';
import {View, Text} from 'react-native';

import PostStyle from './Post.style';

type Props = {
  text: string,
};

export default class Post extends Component<Props> {
  render() {
    const {text} = this.props;
    return (
      <View style={PostStyle.container}>
        <Text>{text}</Text>
      </View>
    );
  }
}

Post.propTypes = {
  text: PropTypes.string,
};