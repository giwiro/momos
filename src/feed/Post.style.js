// @flow
import {StyleSheet} from 'react-native';

export default StyleSheet.create({
  container: {
    backgroundColor: 'white',
    flex: 1,
    padding: 12,
    borderRadius: 4,
    shadowColor: '#000',
    shadowOffset: {
      width: 0,
      height: 2,
    },
    shadowOpacity: 0.25,
    shadowRadius: 3.84,

    elevation: 5,
  },
});
