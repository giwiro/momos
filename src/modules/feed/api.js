// @flow
import {getCreator} from '../../http';

import type {Post} from '../../entities';
import type {RequestOptions} from '../../http';

export type FetchPostApiResponse = {
  data: Post[],
  pagination: {
    total_count: number,
    count: number,
    offset: number,
  },
  meta: {
    status: number,
    msg: string,
    response_id: string,
  },
};

export const endpoints = Object.freeze({
  FETCH_POSTS: 'http://api.giphy.com/v1/stickers/trending',
});

export const fetchPostsApi = (o: RequestOptions) =>
  getCreator(endpoints.FETCH_POSTS, o);
