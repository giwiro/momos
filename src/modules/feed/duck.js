// @flow
import {
  ActionCreator,
  createReducer,
  generateActionCreators,
} from '../../store/utils';
import {ofType} from 'redux-observable';
import {of} from 'rxjs';
import {catchError, map, switchMap} from 'rxjs/operators';
import {fetchPostsApi} from './api';
import {ErrorSet} from '../error/duck';

import type {Action} from '../../store/utils';
import type {ActionsObservable} from 'redux-observable';
import type {Post} from '../../entities';
import type {FetchPostApiResponse} from './api';

export type FeedAction = Action & {
  posts?: Post[],
};

export type FeedState = {
  isFetching?: boolean,
};

export class FeedFetchPosts extends ActionCreator<FeedAction> {}

export class FeedFetchPostsSuccess extends ActionCreator<FeedAction> {}

export const actions = generateActionCreators([
  FeedFetchPosts,
  FeedFetchPostsSuccess,
]);

export const initialState = {};

export default createReducer(initialState, {
  [FeedFetchPosts.type]: (state: FeedState, action: FeedAction) => ({
    isFetching: true,
  }),
  [FeedFetchPostsSuccess]: (state: FeedState, action: FeedAction) => ({
    posts: action.posts,
  }),
});

export const fetchPostsEpic = (action$: ActionsObservable) =>
  action$.pipe(
    ofType(FeedFetchPosts.type),
    switchMap((action: FeedAction) =>
      fetchPostsApi().pipe(
        map(
          (response: FetchPostApiResponse) =>
            new FeedFetchPostsSuccess({posts: response.data}),
        ),
        catchError(error => {
          console.log('error: ', error);
          return of(new ErrorSet());
        }),
      ),
    ),
  );
