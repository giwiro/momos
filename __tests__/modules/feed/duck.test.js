// @flow
import reducer, {
  FeedFetchPosts,
  FeedFetchPostsSuccess,
  FeedFetchPostsError,
  initialState,
  fetchPostsEpic,
} from '../../../src/modules/feed/duck';
import {TestScheduler} from 'rxjs/testing';
import {fetchPostsApi} from '../../../src/modules/feed/api';

jest.mock('../../../src/modules/feed/api');

const mockApiResponse = {
  data: [
    {
      type: 'gif',
      id: '2fPig3PMSsZqIbq0Tc',
      url:
        'https:\\/\\/giphy.com\\/stickers\\/thanks-thank-you-2fPig3PMSsZqIbq0Tc',
      slug: 'thanks-thank-you-2fPig3PMSsZqIbq0Tc',
      bitly_gif_url: 'https:\\/\\/gph.is\\/2mmhfxf',
      bitly_url: 'https:\\/\\/gph.is\\/2mmhfxf',
      embed_url: 'https:\\/\\/giphy.com\\/embed\\/2fPig3PMSsZqIbq0Tc',
      username: 'carawrrr',
      source: 'www.carawrrr.com',
      title: 'Thanks Thank You Sticker by Carawrrr',
      rating: 'g',
      content_url: '',
      source_tld: '',
      source_post_url: 'www.carawrrr.com',
      is_sticker: 1,
      import_datetime: '2018-07-16 15:11:56',
      trending_datetime: '2020-03-25 22:00:28',
      images: {
        hd: {
          height: '1388',
          mp4:
            'https:\\/\\/media1.giphy.com\\/media\\/2fPig3PMSsZqIbq0Tc\\/giphy-hd.mp4?cid=4a60ab043af3077f5a679962db54a09c17ddc5c6b6fa5d7a&rid=giphy-hd.mp4',
          mp4_size: '409144',
          width: '1920',
        },
        downsized_large: {
          height: '347',
          size: '63193',
          url:
            'https:\\/\\/media1.giphy.com\\/media\\/2fPig3PMSsZqIbq0Tc\\/giphy.gif?cid=4a60ab043af3077f5a679962db54a09c17ddc5c6b6fa5d7a&rid=giphy.gif',
          width: '480',
        },
        fixed_height_small_still: {
          height: '100',
          size: '6541',
          url:
            'https:\\/\\/media1.giphy.com\\/media\\/2fPig3PMSsZqIbq0Tc\\/100_s.gif?cid=4a60ab043af3077f5a679962db54a09c17ddc5c6b6fa5d7a&rid=100_s.gif',
          width: '139',
        },
        original: {
          frames: '3',
          hash: '8b4f0894ae3c31235a56fb6ea3ba06e8',
          height: '347',
          mp4:
            'https:\\/\\/media1.giphy.com\\/media\\/2fPig3PMSsZqIbq0Tc\\/giphy.mp4?cid=4a60ab043af3077f5a679962db54a09c17ddc5c6b6fa5d7a&rid=giphy.mp4',
          mp4_size: '65578',
          size: '63193',
          url:
            'https:\\/\\/media1.giphy.com\\/media\\/2fPig3PMSsZqIbq0Tc\\/giphy.gif?cid=4a60ab043af3077f5a679962db54a09c17ddc5c6b6fa5d7a&rid=giphy.gif',
          webp:
            'https:\\/\\/media1.giphy.com\\/media\\/2fPig3PMSsZqIbq0Tc\\/giphy.webp?cid=4a60ab043af3077f5a679962db54a09c17ddc5c6b6fa5d7a&rid=giphy.webp',
          webp_size: '60870',
          width: '480',
        },
        user: {
          avatar_url:
            'https:\\/\\/media3.giphy.com\\/avatars\\/carawrrr\\/nZ9QpWObQIAE.jpg',
          banner_url: '',
          profile_url: 'https:\\/\\/giphy.com\\/carawrrr\\/',
          username: 'carawrrr',
          display_name: 'Carawrrr',
          is_verified: true,
        },
        analytics_response_payload:
          'e=Z2lmX2lkPTJmUGlnM1BNU3NacUlicTBUYyZldmVudF90eXBlPUdJRl9UUkVORElORyZjaWQ9NGE2MGFiMDQzYWYzMDc3ZjVhNjc5OTYyZGI1NGEwOWMxN2RkYzVjNmI2ZmE1ZDdh',
        analytics: {
          onload: {
            url:
              'https:\\/\\/giphy-analytics.giphy.com\\/simple_analytics?response_id=3af3077f5a679962db54a09c17ddc5c6b6fa5d7a&event_type=GIF_TRENDING&gif_id=2fPig3PMSsZqIbq0Tc&action_type=SEEN',
          },
          onclick: {
            url:
              'https:\\/\\/giphy-analytics.giphy.com\\/simple_analytics?response_id=3af3077f5a679962db54a09c17ddc5c6b6fa5d7a&event_type=GIF_TRENDING&gif_id=2fPig3PMSsZqIbq0Tc&action_type=CLICK',
          },
          consent: {
            url:
              'https:\\/\\/giphy-analytics.giphy.com\\/simple_analytics?response_id=3af3077f5a679962db54a09c17ddc5c6b6fa5d7a&event_type=GIF_TRENDING&gif_id=2fPig3PMSsZqIbq0Tc&action_type=SENT',
          },
        },
      },
    },
  ],
  pagination: {
    total_count: 34021,
    count: 30,
    offset: 0,
  },
  meta: {
    status: 200,
    msg: 'OK',
    response_id: '3af3077f5a679962db54a09c17ddc5c6b6fa5d7a',
  },
};

describe('Action creators', () => {
  it('should return an action to fetch posts', () =>
    expect(new FeedFetchPosts()).toMatchSnapshot());

  it('should return an action to fetch posts success', () =>
    expect(
      new FeedFetchPostsSuccess({
        posts: (mockApiResponse.data: any),
        fetchFromTop: true,
      }),
    ).toMatchSnapshot());

  it('should return an action to fetch posts error', () =>
    expect(new FeedFetchPostsError()).toMatchSnapshot());
});

describe('Reducer', () => {
  it('should return initialState', () =>
    expect(initialState).toMatchSnapshot());

  it('should return the state with is loading', () =>
    expect(reducer(initialState, new FeedFetchPosts())).toMatchSnapshot());

  it('should return the state with is loading from top', () =>
    expect(
      reducer(initialState, new FeedFetchPosts({fetchFromTop: true})),
    ).toMatchSnapshot());

  it('should return the state with the posts when there were none', () => {
    jest.spyOn(Date, 'now').mockImplementation(() => 1585202345469);
    expect(
      reducer(
        initialState,
        new FeedFetchPostsSuccess({
          fetchFromTop: true,
          posts: (mockApiResponse.data: any),
        }),
      ),
    ).toMatchSnapshot();
  });

  it('should return the state without loading', () =>
    expect(reducer(initialState, new FeedFetchPostsError())).toMatchSnapshot());
});

describe('Epics', () => {
  let testScheduler: TestScheduler;

  beforeEach(() => {
    testScheduler = new TestScheduler((actual, expected) =>
      expect(actual).toEqual(expected),
    );
  });

  describe('fetchPostsEpic', () => {
    it('should succeed fetching posts', () => {
      testScheduler.run(({hot, cold, expectObservable}) => {
        const action$ = hot('-a|', {
          a: {
            type: FeedFetchPosts.name,
          },
        });
        const state = {
          value: {
            feed: initialState,
          },
        };
        const output$ = fetchPostsEpic(action$, state);
        // $FlowFixMe
        fetchPostsApi.mockImplementation(() =>
          cold('--a|', {
            a: mockApiResponse,
          }),
        );
        expectObservable(output$).toBe('---a|', {
          a: {
            type: 'FeedFetchPostsSuccess',
            posts: mockApiResponse.data,
            fetchFromTop: undefined,
          },
        });
      });
    });
  });
});
