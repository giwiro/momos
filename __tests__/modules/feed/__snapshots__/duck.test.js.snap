// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Action creators should return an action to fetch posts 1`] = `
Object {
  "type": "FeedFetchPosts",
}
`;

exports[`Action creators should return an action to fetch posts error 1`] = `
Object {
  "type": "FeedFetchPostsError",
}
`;

exports[`Action creators should return an action to fetch posts success 1`] = `
Object {
  "fetchFromTop": true,
  "posts": Array [
    Object {
      "bitly_gif_url": "https:\\\\/\\\\/gph.is\\\\/2mmhfxf",
      "bitly_url": "https:\\\\/\\\\/gph.is\\\\/2mmhfxf",
      "content_url": "",
      "embed_url": "https:\\\\/\\\\/giphy.com\\\\/embed\\\\/2fPig3PMSsZqIbq0Tc",
      "id": "2fPig3PMSsZqIbq0Tc",
      "images": Object {
        "analytics": Object {
          "consent": Object {
            "url": "https:\\\\/\\\\/giphy-analytics.giphy.com\\\\/simple_analytics?response_id=3af3077f5a679962db54a09c17ddc5c6b6fa5d7a&event_type=GIF_TRENDING&gif_id=2fPig3PMSsZqIbq0Tc&action_type=SENT",
          },
          "onclick": Object {
            "url": "https:\\\\/\\\\/giphy-analytics.giphy.com\\\\/simple_analytics?response_id=3af3077f5a679962db54a09c17ddc5c6b6fa5d7a&event_type=GIF_TRENDING&gif_id=2fPig3PMSsZqIbq0Tc&action_type=CLICK",
          },
          "onload": Object {
            "url": "https:\\\\/\\\\/giphy-analytics.giphy.com\\\\/simple_analytics?response_id=3af3077f5a679962db54a09c17ddc5c6b6fa5d7a&event_type=GIF_TRENDING&gif_id=2fPig3PMSsZqIbq0Tc&action_type=SEEN",
          },
        },
        "analytics_response_payload": "e=Z2lmX2lkPTJmUGlnM1BNU3NacUlicTBUYyZldmVudF90eXBlPUdJRl9UUkVORElORyZjaWQ9NGE2MGFiMDQzYWYzMDc3ZjVhNjc5OTYyZGI1NGEwOWMxN2RkYzVjNmI2ZmE1ZDdh",
        "downsized_large": Object {
          "height": "347",
          "size": "63193",
          "url": "https:\\\\/\\\\/media1.giphy.com\\\\/media\\\\/2fPig3PMSsZqIbq0Tc\\\\/giphy.gif?cid=4a60ab043af3077f5a679962db54a09c17ddc5c6b6fa5d7a&rid=giphy.gif",
          "width": "480",
        },
        "fixed_height_small_still": Object {
          "height": "100",
          "size": "6541",
          "url": "https:\\\\/\\\\/media1.giphy.com\\\\/media\\\\/2fPig3PMSsZqIbq0Tc\\\\/100_s.gif?cid=4a60ab043af3077f5a679962db54a09c17ddc5c6b6fa5d7a&rid=100_s.gif",
          "width": "139",
        },
        "hd": Object {
          "height": "1388",
          "mp4": "https:\\\\/\\\\/media1.giphy.com\\\\/media\\\\/2fPig3PMSsZqIbq0Tc\\\\/giphy-hd.mp4?cid=4a60ab043af3077f5a679962db54a09c17ddc5c6b6fa5d7a&rid=giphy-hd.mp4",
          "mp4_size": "409144",
          "width": "1920",
        },
        "original": Object {
          "frames": "3",
          "hash": "8b4f0894ae3c31235a56fb6ea3ba06e8",
          "height": "347",
          "mp4": "https:\\\\/\\\\/media1.giphy.com\\\\/media\\\\/2fPig3PMSsZqIbq0Tc\\\\/giphy.mp4?cid=4a60ab043af3077f5a679962db54a09c17ddc5c6b6fa5d7a&rid=giphy.mp4",
          "mp4_size": "65578",
          "size": "63193",
          "url": "https:\\\\/\\\\/media1.giphy.com\\\\/media\\\\/2fPig3PMSsZqIbq0Tc\\\\/giphy.gif?cid=4a60ab043af3077f5a679962db54a09c17ddc5c6b6fa5d7a&rid=giphy.gif",
          "webp": "https:\\\\/\\\\/media1.giphy.com\\\\/media\\\\/2fPig3PMSsZqIbq0Tc\\\\/giphy.webp?cid=4a60ab043af3077f5a679962db54a09c17ddc5c6b6fa5d7a&rid=giphy.webp",
          "webp_size": "60870",
          "width": "480",
        },
        "user": Object {
          "avatar_url": "https:\\\\/\\\\/media3.giphy.com\\\\/avatars\\\\/carawrrr\\\\/nZ9QpWObQIAE.jpg",
          "banner_url": "",
          "display_name": "Carawrrr",
          "is_verified": true,
          "profile_url": "https:\\\\/\\\\/giphy.com\\\\/carawrrr\\\\/",
          "username": "carawrrr",
        },
      },
      "import_datetime": "2018-07-16 15:11:56",
      "is_sticker": 1,
      "rating": "g",
      "slug": "thanks-thank-you-2fPig3PMSsZqIbq0Tc",
      "source": "www.carawrrr.com",
      "source_post_url": "www.carawrrr.com",
      "source_tld": "",
      "title": "Thanks Thank You Sticker by Carawrrr",
      "trending_datetime": "2020-03-25 22:00:28",
      "type": "gif",
      "url": "https:\\\\/\\\\/giphy.com\\\\/stickers\\\\/thanks-thank-you-2fPig3PMSsZqIbq0Tc",
      "username": "carawrrr",
    },
  ],
  "type": "FeedFetchPostsSuccess",
}
`;

exports[`Reducer should return initialState 1`] = `
Object {
  "posts": Immutable.OrderedMap {},
}
`;

exports[`Reducer should return the state with is loading 1`] = `
Object {
  "isFetching": true,
  "isFetchingFromTop": false,
  "posts": Immutable.OrderedMap {},
}
`;

exports[`Reducer should return the state with is loading from top 1`] = `
Object {
  "isFetching": true,
  "isFetchingFromTop": true,
  "posts": Immutable.OrderedMap {},
}
`;

exports[`Reducer should return the state with the posts when there were none 1`] = `
Object {
  "isFetching": false,
  "isFetchingFromTop": false,
  "lastRequestDate": 1585202345469,
  "posts": Immutable.OrderedMap {
    "2fPig3PMSsZqIbq0Tc": Object {
      "bitly_gif_url": "https:\\\\/\\\\/gph.is\\\\/2mmhfxf",
      "bitly_url": "https:\\\\/\\\\/gph.is\\\\/2mmhfxf",
      "content_url": "",
      "embed_url": "https:\\\\/\\\\/giphy.com\\\\/embed\\\\/2fPig3PMSsZqIbq0Tc",
      "id": "2fPig3PMSsZqIbq0Tc",
      "images": Object {
        "analytics": Object {
          "consent": Object {
            "url": "https:\\\\/\\\\/giphy-analytics.giphy.com\\\\/simple_analytics?response_id=3af3077f5a679962db54a09c17ddc5c6b6fa5d7a&event_type=GIF_TRENDING&gif_id=2fPig3PMSsZqIbq0Tc&action_type=SENT",
          },
          "onclick": Object {
            "url": "https:\\\\/\\\\/giphy-analytics.giphy.com\\\\/simple_analytics?response_id=3af3077f5a679962db54a09c17ddc5c6b6fa5d7a&event_type=GIF_TRENDING&gif_id=2fPig3PMSsZqIbq0Tc&action_type=CLICK",
          },
          "onload": Object {
            "url": "https:\\\\/\\\\/giphy-analytics.giphy.com\\\\/simple_analytics?response_id=3af3077f5a679962db54a09c17ddc5c6b6fa5d7a&event_type=GIF_TRENDING&gif_id=2fPig3PMSsZqIbq0Tc&action_type=SEEN",
          },
        },
        "analytics_response_payload": "e=Z2lmX2lkPTJmUGlnM1BNU3NacUlicTBUYyZldmVudF90eXBlPUdJRl9UUkVORElORyZjaWQ9NGE2MGFiMDQzYWYzMDc3ZjVhNjc5OTYyZGI1NGEwOWMxN2RkYzVjNmI2ZmE1ZDdh",
        "downsized_large": Object {
          "height": "347",
          "size": "63193",
          "url": "https:\\\\/\\\\/media1.giphy.com\\\\/media\\\\/2fPig3PMSsZqIbq0Tc\\\\/giphy.gif?cid=4a60ab043af3077f5a679962db54a09c17ddc5c6b6fa5d7a&rid=giphy.gif",
          "width": "480",
        },
        "fixed_height_small_still": Object {
          "height": "100",
          "size": "6541",
          "url": "https:\\\\/\\\\/media1.giphy.com\\\\/media\\\\/2fPig3PMSsZqIbq0Tc\\\\/100_s.gif?cid=4a60ab043af3077f5a679962db54a09c17ddc5c6b6fa5d7a&rid=100_s.gif",
          "width": "139",
        },
        "hd": Object {
          "height": "1388",
          "mp4": "https:\\\\/\\\\/media1.giphy.com\\\\/media\\\\/2fPig3PMSsZqIbq0Tc\\\\/giphy-hd.mp4?cid=4a60ab043af3077f5a679962db54a09c17ddc5c6b6fa5d7a&rid=giphy-hd.mp4",
          "mp4_size": "409144",
          "width": "1920",
        },
        "original": Object {
          "frames": "3",
          "hash": "8b4f0894ae3c31235a56fb6ea3ba06e8",
          "height": "347",
          "mp4": "https:\\\\/\\\\/media1.giphy.com\\\\/media\\\\/2fPig3PMSsZqIbq0Tc\\\\/giphy.mp4?cid=4a60ab043af3077f5a679962db54a09c17ddc5c6b6fa5d7a&rid=giphy.mp4",
          "mp4_size": "65578",
          "size": "63193",
          "url": "https:\\\\/\\\\/media1.giphy.com\\\\/media\\\\/2fPig3PMSsZqIbq0Tc\\\\/giphy.gif?cid=4a60ab043af3077f5a679962db54a09c17ddc5c6b6fa5d7a&rid=giphy.gif",
          "webp": "https:\\\\/\\\\/media1.giphy.com\\\\/media\\\\/2fPig3PMSsZqIbq0Tc\\\\/giphy.webp?cid=4a60ab043af3077f5a679962db54a09c17ddc5c6b6fa5d7a&rid=giphy.webp",
          "webp_size": "60870",
          "width": "480",
        },
        "user": Object {
          "avatar_url": "https:\\\\/\\\\/media3.giphy.com\\\\/avatars\\\\/carawrrr\\\\/nZ9QpWObQIAE.jpg",
          "banner_url": "",
          "display_name": "Carawrrr",
          "is_verified": true,
          "profile_url": "https:\\\\/\\\\/giphy.com\\\\/carawrrr\\\\/",
          "username": "carawrrr",
        },
      },
      "import_datetime": "2018-07-16 15:11:56",
      "is_sticker": 1,
      "rating": "g",
      "slug": "thanks-thank-you-2fPig3PMSsZqIbq0Tc",
      "source": "www.carawrrr.com",
      "source_post_url": "www.carawrrr.com",
      "source_tld": "",
      "title": "Thanks Thank You Sticker by Carawrrr",
      "trending_datetime": "2020-03-25 22:00:28",
      "type": "gif",
      "url": "https:\\\\/\\\\/giphy.com\\\\/stickers\\\\/thanks-thank-you-2fPig3PMSsZqIbq0Tc",
      "username": "carawrrr",
    },
  },
}
`;

exports[`Reducer should return the state without loading 1`] = `
Object {
  "isFetching": false,
  "isFetchingFromTop": false,
  "posts": Immutable.OrderedMap {},
}
`;
